src/core/channel/client_uchannel.c

```diff
@@ -0,0 +1,573 @@
+  GPR_ASSERT(op->bind_pollset == NULL);
+
+  if (op->on_connectivity_state_change != NULL) {
+    grpc_connectivity_state_notify_on_state_change(
+        exec_ctx, &chand->state_tracker, op->connectivity_state,
+        op->on_connectivity_state_change);
+    op->on_connectivity_state_change = NULL;
+    op->connectivity_state = NULL;
+  }
+
+  if (op->disconnect) {
+    grpc_connectivity_state_set(exec_ctx, &chand->state_tracker,
+                                GRPC_CHANNEL_FATAL_FAILURE, "disconnect");
+  }
+}
+
+/* Constructor for call_data */
+static void cmc_init_call_elem(grpc_exec_ctx *exec_ctx, grpc_call_element *elem,
+                               const void *server_transport_data,
+                               grpc_transport_stream_op *initial_op) {
+  call_data *calld = elem->call_data;
+  memset(calld, 0, sizeof(call_data));
+
+  /* TODO(ctiller): is there something useful we can do here? */
+  GPR_ASSERT(initial_op == NULL);
+
+  GPR_ASSERT(elem->filter == &grpc_client_uchannel_filter);
+  GPR_ASSERT(server_transport_data == NULL);
+  gpr_mu_init(&calld->mu_state);
+  calld->elem = elem;
+  calld->state = CALL_CREATED;
+  calld->deadline = gpr_inf_future(GPR_CLOCK_REALTIME);
+}
+
+/* Destructor for call_data */
+static void cmc_destroy_call_elem(grpc_exec_ctx *exec_ctx,
+                                  grpc_call_element *elem) {
+  call_data *calld = elem->call_data;
+  grpc_subchannel_call *subchannel_call;
+
+  /* if the call got activated, we need to destroy the child stack also, and
+     remove it from the in-flight requests tracked by the child_entry we
+     picked */
+  gpr_mu_lock(&calld->mu_state);
+  switch (calld->state) {
+    case CALL_ACTIVE:
+      subchannel_call = calld->subchannel_call;
+      gpr_mu_unlock(&calld->mu_state);
+      GRPC_SUBCHANNEL_CALL_UNREF(exec_ctx, subchannel_call, "client_uchannel");
+      break;
+    case CALL_CREATED:
+    case CALL_CANCELLED:
+      gpr_mu_unlock(&calld->mu_state);
+      break;
+    case CALL_WAITING_FOR_CALL:
+    case CALL_WAITING_FOR_SEND:
+      gpr_log(GPR_ERROR, "should never reach here");
+      abort();
+      break;
+  }
+}
+
+/* Constructor for channel_data */
+static void cmc_init_channel_elem(grpc_exec_ctx *exec_ctx,
+                                  grpc_channel_element *elem,
+                                  grpc_channel *master,
+                                  const grpc_channel_args *args,
+                                  grpc_mdctx *metadata_context, int is_first,
+                                  int is_last) {
+  size_t i;
+  int subchannel_pointer_arg_found = 0;
+  channel_data *chand = elem->channel_data;
+
+  memset(chand, 0, sizeof(*chand));
+
+  grpc_closure_init(&chand->connectivity_cb, monitor_subchannel, chand);
+  GPR_ASSERT(is_last);
+  GPR_ASSERT(elem->filter == &grpc_client_uchannel_filter);
+
+  chand->mdctx = metadata_context;
+  chand->master = master;
+  for (i = 0; i < args->num_args; i++) {
```

Comment:
Can we drop the GRPC_MICROCHANNEL_SUBCHANNEL_ARG, and instead have a post-creation initializer (grpc_microchannel_set_subchannel())